File: backend/app/routers/payments.py
"""Stripe payments — the real authorize/capture that replaced the app's
simulated `setTimeout` in src/lib/placeOrder.ts.

Flow (the app drives it in this order):

  1. App generates the order id up front and POSTs the cart here as **item ids
     and quantities only** — no prices, no total.
  2. This module reprices the cart from the backend's own menu (`pricing.py`)
     and creates a PaymentIntent for that amount, stamping the breakdown and
     the order id into the intent's metadata.
  3. App presents Stripe's PaymentSheet with the returned client secret.
  4. On success the app calls POST /orders with the same order id and the
     payment intent id; `routers/orders.py` re-reads the intent from Stripe,
     confirms it actually succeeded, and takes the recorded amounts from the
     intent metadata rather than from the request body.

So the client never gets to name a price, and an order can't exist without a
matching succeeded charge. The `order_id` doubles as the Stripe idempotency
key, so a double tap reuses one intent instead of creating a second charge.
"""

import uuid

import stripe
from fastapi import APIRouter, Header, HTTPException, Request
from pydantic import BaseModel, Field

from .. import config, pricing
from ..postgrest import REST_URL, bearer_from, client, headers, jwt_sub
from ..rate_limit import rate_limited

router = APIRouter(prefix="/payments")


def _require_stripe() -> None:
    if not config.STRIPE_CONFIGURED:
        raise HTTPException(503, "Payments are not configured on the backend")
    stripe.api_key = config.STRIPE_SECRET_KEY


class QuoteItem(BaseModel):
    # The menu item id from src/data/menu.ts — NOT a display name, and never a
    # price. Prices come from pricing.PRICES and nowhere else.
    id: str = Field(min_length=1, max_length=200)
    quantity: int = Field(ge=1, le=pricing.MAX_QUANTITY)


class IntentBody(BaseModel):
    # Client-generated so the app can reference the order before it's paid for,
    # and so a retry lands on the same Stripe idempotency key.
    order_id: str = Field(min_length=1, max_length=64)
    items: list[QuoteItem] = Field(min_length=1)
    tip_cents: int = Field(default=0, ge=0, le=pricing.MAX_TIP_CENTS)


@router.post("/intent")
async def create_intent(body: IntentBody, authorization: str | None = Header(default=None)):
    _require_stripe()

    # Creating PaymentIntents is cheap but not free, and it's an unauthenticated
    # endpoint (checkout works in demo mode without a Supabase session).
    if rate_limited(f"payments:intent:{body.order_id}"):
        raise HTTPException(429, "Too many payment attempts — wait a moment and try again.")

    try:
        uuid.UUID(body.order_id)
    except ValueError:
        raise HTTPException(422, "order_id must be a UUID") from None

    try:
        quote = pricing.quote([(i.id, i.quantity) for i in body.items], body.tip_cents)
    except pricing.PricingError as exc:
        # 422 rather than 400: the request was well-formed, the cart wasn't
        # priceable. Same status the delivery-location gate uses.
        raise HTTPException(422, str(exc)) from None

    # Ties the charge to the signed-in customer when there is one. Optional:
    # demo mode has no Supabase session and still needs to check out.
    customer_id = jwt_sub(bearer_from(authorization)) or ""

    metadata = {
        **quote.as_metadata(),
        "order_id": body.order_id,
        "customer_id": customer_id,
        "app": "localgo-consumer",
    }

    try:
        intent = stripe.PaymentIntent.create(
            amount=quote.total_cents,
            currency=quote.currency,
            metadata=metadata,
            automatic_payment_methods={"enabled": True},
            # Same order id => same intent. Protects against double-taps and
            # retries after a dropped response.
            idempotency_key=f"localgo-order-{body.order_id}",
        )
    except stripe.StripeError as exc:
        print(f"[payments] intent create failed: {type(exc).__name__}: {exc}")
        raise HTTPException(502, "Could not start the payment. Please try again.") from None

    # An idempotent replay can return an intent whose amount was fixed by the
    # first call. If the cart changed in between, the client would present a
    # sheet for the old amount — refuse instead of silently charging the wrong
    # total.
    if intent.amount != quote.total_cents:
        raise HTTPException(
            409,
            "This order was already started at a different total. Start a new checkout.",
        )

    return {
        "publishable_key": config.STRIPE_PUBLISHABLE_KEY,
        "client_secret": intent.client_secret,
        "payment_intent_id": intent.id,
        "currency": quote.currency,
        "subtotal_cents": quote.subtotal_cents,
        "delivery_fee_cents": quote.delivery_fee_cents,
        "tip_cents": quote.tip_cents,
        "total_cents": quote.total_cents,
    }


async def _order_exists(order_id: str) -> bool | None:
    """Has this order actually been recorded? None means we couldn't tell.

    Needs the service-role key: the anon role has INSERT but no SELECT on
    `orders` (the RLS gotcha documented in routers/orders.py), so an anon read
    here would come back empty and report every charge as orphaned.
    """
    if not config.SUPABASE_CONFIGURED or not config.SUPABASE_SERVICE_ROLE_KEY:
        return None
    try:
        res = await client.get(
            f"{REST_URL}/orders",
            params={"id": f"eq.{order_id}", "select": "id"},
            headers=headers(service=True),
        )
    except Exception as exc:  # noqa: BLE001
        print(f"[payments] order lookup failed: {type(exc).__name__}: {exc}")
        return None
    if res.status_code != 200:
        print(f"[payments] order lookup returned {res.status_code}")
        return None
    return len(res.json()) > 0


@router.post("/webhook")
async def stripe_webhook(request: Request, stripe_signature: str = Header(default="")):
    """Stripe's out-of-band truth about a charge.

    The happy path doesn't need this — POST /orders verifies the intent
    directly. It matters for the case the app can't report: the customer's
    payment succeeds but the phone dies, loses signal, or the app is killed
    before it calls /orders. Those charges would otherwise be money taken for an
    order no driver ever sees, with nothing anywhere recording it.

    On a successful payment this looks up the matching order row and logs an
    ORPHANED CHARGE line when there isn't one, so it can be refunded or
    completed by hand.

    RACE: Stripe often delivers this before the app's own POST /orders lands, so
    a freshly orphaned charge can be a false alarm that resolves itself seconds
    later. Always re-check the order id before refunding anything. Catching the
    genuine cases properly needs a periodic sweep over recent intents, not just
    this hook — see LIST.md.
    """
    if not config.STRIPE_WEBHOOK_SECRET:
        raise HTTPException(503, "Webhooks are not configured")

    payload = await request.body()
    try:
        event = stripe.Webhook.construct_event(
            payload, stripe_signature, config.STRIPE_WEBHOOK_SECRET
        )
    except (ValueError, stripe.SignatureVerificationError) as exc:
        # Unsigned or forged — never act on it.
        print(f"[payments] webhook rejected: {type(exc).__name__}: {exc}")
        raise HTTPException(400, "Invalid webhook signature") from None

    kind = event["type"]
    # StripeObject, not a dict — attribute access with a default, never .get().
    obj = event["data"]["object"]

    if kind == "payment_intent.succeeded":
        metadata = getattr(obj, "metadata", None)
        order_id = getattr(metadata, "order_id", None) if metadata is not None else None
        intent_id = getattr(obj, "id", None)
        amount = getattr(obj, "amount", None)
        print(f"[payments] succeeded intent={intent_id} order={order_id} amount={amount}")

        if order_id:
            recorded = await _order_exists(order_id)
            if recorded is False:
                # Grep for "ORPHANED CHARGE" in the Render logs.
                print(
                    f"[payments] ORPHANED CHARGE intent={intent_id} order={order_id} "
                    f"amount={amount} — money taken with no order row. Re-check before "
                    f"refunding; the app's POST /orders may still be in flight."
                )
    elif kind == "payment_intent.payment_failed":
        last_error = getattr(obj, "last_payment_error", None)
        reason = getattr(last_error, "message", "unknown") if last_error is not None else "unknown"
        print(f"[payments] failed intent={getattr(obj, 'id', None)} reason={reason}")

    return {"received": True}

File: src/lib/payments.ts
// Stripe PaymentSheet — the real charge that replaced checkout's simulated
// `setTimeout`. Deliberately a plain module rather than a hook: both entry
// points to checkout (the Cart screen and Goer's in-chat confirmation card) go
// through `placeOrder`, which is called outside React, and
// initPaymentSheet/presentPaymentSheet are module-level functions.
//
// The app never computes or sends an amount. It posts item ids and quantities;
// the backend reprices from its own menu (backend/app/pricing.py) and returns a
// client secret for the amount it decided. Whatever total the cart is showing
// is a display value only.
//
// The publishable key also comes from that response rather than from a bundled
// env var, so there is exactly one place a Stripe key is configured (the
// backend) and the two can never drift apart.
import {
  initPaymentSheet,
  initStripe,
  presentPaymentSheet,
} from '@stripe/stripe-react-native';
import type { CartLine } from '../types';
import { API_URL, isBackendConfigured } from './api';

/** Real card payments require the backend that issues the PaymentIntent. */
export const isPaymentsConfigured = isBackendConfigured;

export interface PaymentQuote {
  paymentIntentId: string;
  clientSecret: string;
  publishableKey: string;
  currency: string;
  subtotalCents: number;
  deliveryFeeCents: number;
  tipCents: number;
  /** What the customer will actually be charged, per the server. */
  totalCents: number;
}

export type PayOutcome =
  | { status: 'succeeded'; paymentIntentId: string; quote: PaymentQuote }
  /** User dismissed the sheet. Not an error — the caller should stay put and say nothing. */
  | { status: 'canceled' }
  | { status: 'failed'; message: string };

let stripeReadyFor: string | null = null;

/** initStripe is idempotent but not free; only re-run it if the key changes. */
async function ensureStripe(publishableKey: string): Promise<void> {
  if (stripeReadyFor === publishableKey) return;
  await initStripe({ publishableKey, merchantIdentifier: 'merchant.com.samerodeh.localgo' });
  stripeReadyFor = publishableKey;
}

/**
 * Ask the backend to price this cart and open a PaymentIntent for it.
 * Throws on any failure — the message is safe to show the user.
 */
export async function createPaymentIntent(input: {
  orderId: string;
  lines: CartLine[];
  tip: number;
  accessToken: string | null;
}): Promise<PaymentQuote> {
  if (!API_URL) throw new Error('Payments are unavailable right now.');

  const res = await fetch(`${API_URL}/payments/intent`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      ...(input.accessToken ? { Authorization: `Bearer ${input.accessToken}` } : {}),
    },
    body: JSON.stringify({
      order_id: input.orderId,
      // Ids and quantities only. Prices are the server's business.
      items: input.lines.map((l) => ({ id: l.item.id, quantity: l.quantity })),
      tip_cents: Math.round(input.tip * 100),
    }),
  });

  if (!res.ok) {
    // The backend's 422s are written for humans (unknown item, tip too large,
    // menu unavailable), so surface them as-is when we can parse one out.
    const detail = await res
      .json()
      .then((body) => (typeof body?.detail === 'string' ? body.detail : null))
      .catch(() => null);
    throw new Error(detail ?? 'We could not start your payment. Please try again.');
  }

  const data = await res.json();
  return {
    paymentIntentId: data.payment_intent_id,
    clientSecret: data.client_secret,
    publishableKey: data.publishable_key,
    currency: data.currency,
    subtotalCents: data.subtotal_cents,
    deliveryFeeCents: data.delivery_fee_cents,
    tipCents: data.tip_cents,
    totalCents: data.total_cents,
  };
}

/**
 * Full charge: quote it, open the sheet, wait for the customer.
 *
 * Never throws — every failure comes back as a `PayOutcome` so the caller can
 * tell "user changed their mind" apart from "card declined", which matter
 * differently to the UI.
 */
export async function payForOrder(input: {
  orderId: string;
  lines: CartLine[];
  tip: number;
  accessToken: string | null;
}): Promise<PayOutcome> {
  let quote: PaymentQuote;
  try {
    quote = await createPaymentIntent(input);
  } catch (err) {
    return {
      status: 'failed',
      message: err instanceof Error ? err.message : 'We could not start your payment.',
    };
  }

  try {
    await ensureStripe(quote.publishableKey);

    const init = await initPaymentSheet({
      merchantDisplayName: 'LocalGO',
      paymentIntentClientSecret: quote.clientSecret,
      // Must match `expo.scheme` in app.json, or bank redirect flows (3-D
      // Secure) can't get back into the app and the payment hangs.
      returnURL: 'localgo://stripe-redirect',
      // Delivery starts as soon as the order lands on the driver feed, so we
      // can't accept methods that settle days later.
      allowsDelayedPaymentMethods: false,
    });
    if (init.error) {
      return { status: 'failed', message: init.error.message };
    }

    const result = await presentPaymentSheet();
    if (result.error) {
      // Stripe reports a dismissed sheet as an error with code 'Canceled'.
      if (result.error.code === 'Canceled') return { status: 'canceled' };
      return { status: 'failed', message: result.error.message };
    }

    return { status: 'succeeded', paymentIntentId: quote.paymentIntentId, quote };
  } catch (err) {
    return {
      status: 'failed',
      message: err instanceof Error ? err.message : 'Your payment could not be completed.',
    };
  }
}
